xarray/indexes/range_index.py

```diff
@@ -0,0 +1,352 @@
+    def size(self) -> int:
+        return self.dim_size[self.dim]
+
+    @property
+    def step(self) -> float:
+        return (self.stop - self.start) / self.size
+
+    def forward(self, dim_positions: dict[str, Any]) -> dict[Hashable, Any]:
+        positions = dim_positions[self.dim]
+        labels = self.start + positions * self.step
+        return {self.coord_name: labels}
+
+    def reverse(self, coord_labels: dict[Hashable, Any]) -> dict[str, Any]:
+        labels = coord_labels[self.coord_name]
+        positions = (labels - self.start) / self.step
+        return {self.dim: positions}
+
+    def equals(self, other: CoordinateTransform) -> bool:
+        if not isinstance(other, RangeCoordinateTransform):
+            return False
+
+        return (
+            self.start == other.start
+            and self.stop == other.stop
+            and self.size == other.size
+        )
+
+    def slice(self, sl: slice) -> "RangeCoordinateTransform":
+        sl = normalize_slice(sl, self.size)
+
+        # TODO: support reverse transform (i.e., start > stop)?
+        assert sl.start < sl.stop
+
+        new_size = (sl.stop - sl.start) // sl.step
+        new_start = self.start + sl.start * self.step
+        new_stop = new_start + new_size * sl.step * self.step
+
+        return type(self)(
+            new_start, new_stop, new_size, self.coord_name, self.dim, dtype=self.dtype
+        )
+
+
+class RangeIndex(CoordinateTransformIndex):
+    """Xarray index implementing a simple bounded 1-dimension interval with
+    evenly spaced, monotonic floating-point values.
+
+    This index is memory-saving, i.e., the values of its associated coordinate
+    variable are not materialized in memory.
+
+    Do not use :py:meth:`~xarray.indexes.RangeIndex.__init__` directly. Instead
+    use :py:meth:`~xarray.indexes.RangeIndex.arange` or
+    :py:meth:`~xarray.indexes.RangeIndex.linspace`, which are similar to
+    :py:func:`numpy.arange` and :py:func:`numpy.linspace`.
+
+    In the case of a monotonic integer range, it is better using a
+    :py:class:`~xarray.indexes.PandasIndex` that wraps a
+    :py:class:`pandas.RangeIndex`.
+
+    """
+
+    transform: RangeCoordinateTransform
+
+    def __init__(self, transform: RangeCoordinateTransform):
+        super().__init__(transform)
+
+    @classmethod
+    def arange(
+        cls,
+        coord_name: Hashable,
+        dim: str,
+        start: float = 0.0,
+        stop: float = 1.0,
+        step: float = 1.0,
+        dtype: Any = None,
+    ) -> "RangeIndex":
+        """Create a new RangeIndex from given start, stop and step values.
+
+        Parameters
+        ----------
+        coord_name : Hashable
+            Name of the (lazy) coordinate variable that will be created and
+            associated with the new index.
+        dim : str
+            Dimension name.
+        start : float, optional
+            Start of interval (default: 0.0). The interval includes this value.
```

Comment:
Could consider adding a `closed` kwarg like pd.Interval, but in a future PR of course.